### PR TITLE
bounding boxes coordinate fixing in object detection augmentor

### PR DESCRIPTION
**Pull request by lm-jira**
Some augmentors edit the images without modifying the coordinate of bounding boxes (gt_boxes variables), which is reported by [this issue](https://github.com/blue-oil/blueoil/issues/60).
This pull request changes the coordinate of bounding boxes for those augmentors.

## Motivation and Context
Editing images without changing coordinate of bounding boxes results in incorrect data label for training data and affects the accuracy of object detection model

## Description
Edit those classes (only `Pad` in this step) to accept `gt_boxes` as an input,
modify the coordinate of `gt_boxes` according to the way images are padded,
and return the correct information of `gt_boxes`.

## How has this been tested?
I tested it using the same setting as Blueoil's object detection tutorial, which is detecting human face and hand using a part of OpenImagesV4 dataset.
I implemented on server25.

## Screenshots (if appropriate):
The example of bounding boxes in augmented image is as follows.
screen shot 2019-01-11 at 16 43 47

## Types of changes
- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by ruimashita** [comment]
@lm-jira Nice!! Please resolve conflicts

**Comment by ruimashita** [comment]
@lm-jira please implement unit test l.

**Comment by lm-jira** [comment]
@ruimashita  I edited the code as follows

- **Pad**: commit 05898c0157f312cc791808a08a828cb42fa111da
The coordinate of bounding box is added by the padding size as suggested by this [issue](https://github.com/blue-oil/blueoil/issues/60).

- **Crop**: commit d4a294046c3d21e5595b75622277b1a738754017
The bounding box is removed if it is outside the cropped area.
Then, the coordinate is adjusted according to the cropped position.
Finally, if a part of the bounding box is outside the cropped image, the coordinate is changes so that it is at the margin of the image.  (e.g. if x < 0, update x to be 0)

- **Rotate**: commit be361963a49ac3bc871a45ced4d7a3cc5c6c0e22
I used this [article](https://blog.paperspace.com/data-augmentation-for-object-detection-rotation-and-shearing/) as a reference for moving the bounding box.
rotate_box
The bounding box is changed with the following step.
     1. Find the coordinate of four corner points of the bounding boxes.
     2. Rotate the four points using the image's center as the center of rotating.
     3. Find the smallest bounding box that the rotated points in the previous step can fit in (the black box in the rightmost image) and update the corresponding bounding box to that smallest box.

**Comment by iizukak** [comment]
@ruimashita @nlpng Please review this PR.

**Comment by ruimashita** [comment]
@lm-jira 
How much effect to final result the rotate way ?

**Comment by ruimashita** [comment]
@lm-jira 
Could you implement unit test or take screenshots when the changed box is outside of image of pad and crop.
Looks not considered or wrong when outside for me.

**Comment by ruimashita** [comment]
This is not bug, new feature, I think

**Comment by lm-jira** [comment]
> @lm-jira
> How much effect to final result the rotate way ?

Rotating bounding box results in higher overall accuracy.
(Red line = Accuracy of model after rotating bounding box, 
Pink line = Accuracy of model before rotating the bounding box)
screen shot 2019-02-18 at 9 11 18

**Comment by ruimashita** [comment]
@lm-jira
Thanks 😀
But this is not final report. This is middle progress.

**Comment by lm-jira** [comment]
@ruimashita 

> @lm-jira
> Could you implement unit test or take screenshots when the changed box is outside of image of pad and crop.
> Looks not considered or wrong when outside for me.

Do you mean bounding boxes in the images below? 
(cropping image and cut the bounding box to the margin before padding image)
screen shot 2019-02-18 at 16 42 23


> @lm-jira
> Thanks 😀
> But this is not final report. This is middle progress.

Should I write the report comparing the results in Kibela?? 🤔

**Comment by lm-jira** [comment]
@ruimashita 
I write a report comparing the result before and after correcting the `rotate` augmentor [here](https://leapmind.kibe.la/@jira/799).
This time, I increased the number of epochs from 100 to 600 to be able to clearly see the difference.

**Comment by ruimashita** [comment]
@lm-jira 
Thanks. Nice report. But the `MAP` is low, it is not suitable to actual project, I still wonder the method is useful or not.
Could you compare `Rotate` or not using this config ?
https://github.com/blue-oil/blueoil/blob/master/lmnet/configs/core/object_detection/lm_fyolo_quantize_pascalvoc_2007_2012.py

**Comment by lm-jira** [comment]
@ruimashita 
I did the comparison experiment of rotate augmenter using `pascalvoc_2007` as you mentioned.
The difference between before and after modification is very little as shown below.
More detail can be found [here](https://leapmind.kibe.la/@jira/871).

Screen Shot 2019-03-11 at 14 55 27

I also simply search the article for rotating bounding box in object detection model but cannot find.
I'm thinking that rotating the bounding boxes does not worth the increasing performance so I remove the rotate part from this PR and move it to another branch (may be not create a PR for it).

**Comment by a-hanamoto** [comment]
@lm-jira How is the situation of this issue? Do you need any help?

**Comment by lm-jira** [comment]
@a-hanamoto I think the fix of bounding box coordinate for pad and crop augmentation is fine but it doesn't pass the review so I cannot merge it yet.
Because the  Blueoil has a lot of update, I think it'd be better to make a new PR for this issue.

**Comment by a-hanamoto** [comment]
@lm-jira Got it. Thanks! Then could you please make a new PR?

**Comment by lm-jira** [comment]
@a-hanamoto Okay. I will do it when I have time.
